### ISSUE TITLE
CS0612 in generated code cannot be supressed in editorconfig

### ISSUE THREAD

**Issue by hankovich** [dotnet/roslyn#69298]
**Version Used**: .NET 8 preview 6

**Steps to Reproduce**:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Riok.Mapperly" Version="2.8.0" />
  </ItemGroup>

</Project>
```

```csharp
using Riok.Mapperly.Abstractions;

Console.WriteLine(new Source().MapToTarget());

internal sealed class Source
{
    [Obsolete]
    public int Id { get; set; }
}

internal sealed class Target
{
    public int Id { get; set; }
}

[Mapper]
internal static partial class Mapper
{
    public static partial Target MapToTarget(this Source source);
}
```

```ini
root = true

[*.cs]

# CS0612: Member is obsolete
dotnet_diagnostic.CS0612.severity = none
```

This code id generated by `Riok.Mapperly`

```csharp
#nullable enable
internal static partial class Mapper
{
    public static partial global::Target MapToTarget(this global::Source source)
    {
        var target = new global::Target();
        target.Id = source.Id; // <- a diagnostic is reported here
        return target;
    }
}
```

**Diagnostic Id**: CS0612 

**Expected Behavior**:

No CS0612 diagnostic is reported.

**Actual Behavior**:

A diagnostic `C:\Users\Hello\source\repos\ConsoleApp2\ConsoleApp2\Riok.Mapperly\Riok.Mapperly.MapperGenerator\Mapper.g.cs(7,21,7,30): warning CS0612: 'Source.Id' is obsolete` is reported.

**Comment by sharwell** [comment]
Does it work to put `[Obsolete]` on the partial definition?

```cs
[Mapper]
internal static partial class Mapper
{
    [Obsolete]
    public static partial Target MapToTarget(this Source source);
}
```

**Comment by hankovich** [comment]
Yes, the warning is gone.

**Comment by sharwell** [comment]
This issue is technically a duplicate of #47384 (.editorconfig does not apply to source generated files because source generated files are not considered part of the directory tree on disk).

**Comment by sharwell** [comment]
Also related to https://github.com/riok/mapperly/issues/284

I believe the approach I would most recommend is having Riok.Mapperly emit a `#pragma warning suppress CS0618` around the property access for the specific case where `MapperIgnoreObsoleteMembers` has been explicitly provided. In other words, if the code has been explicitly instructed to access an obsolete property during the mapping process, it should suppress warnings resulting from that access.

**Comment by jaredpar** [comment]
Closing as the dupe

**Comment by latonz** [comment]
@sharwell I‘m looking into the [related Mapperly issue](https://github.com/riok/mapperly/issues/610). I had in mind that compiler warnings are suppressed for generated files (with a `<auto-generated />` comment). I couldn‘t find any documentation on this topic. Are there diagnostics which are suppressed for generated files? Is there any documentation available on this topic?

**Comment by sharwell** [comment]
> I had in mind that compiler warnings are suppressed for generated files (with a `<auto-generated />` comment).

Compiler warnings are not affected by an `<auto-generated/>` comment. Analyzer warnings _may_ be affected by such a comment if they opt-in to this behavior (by calling [`ConfigureGeneratedCodeAnalysis`](https://github.com/dotnet/roslyn/blob/66f37d3150ae4082b3170523b75dd823ad023132/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs#L220) without specifying [`GeneratedCodeAnalysisFlags.ReportDiagnostics`](https://github.com/dotnet/roslyn/blob/66f37d3150ae4082b3170523b75dd823ad023132/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs#L282-L286)). 

My recommendation for this issue is to suppress CS0618 only around specific property accesses, and not for the entire file. In addition, this should only occur if the user has explicitly used `MapperIgnoreObsoleteMembers`.

**Comment by sharwell** [comment]
Duplicate of #47384

**Comment by latonz** [comment]
@sharwell thank you for these explanations, we'll look into it.

**Comment by rabberbock** [comment]
An interesting approach that worked for me was to suppress all warnings in a [`.globalconfig`](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-files#global-analyzerconfig), which includes generated files. Then you can override it in the .editorconfig and set it back to warning.
